lib/ruby_indexer/lib/ruby_indexer/index.rb

```diff
@@ -356,6 +356,36 @@ def linearized_ancestors_of(name)
       ancestors
     end
 
+    # Synchronizes a change made to the given indexable path. This method will ensure that new declarations are indexed,
+    # removed declarations removed and that the ancestor linearization cache is cleared if necessary
+    sig { params(indexable: IndexablePath).void }
+    def handle_change(indexable)
+      original_entries = @files_to_entries[indexable.full_path]
+
+      delete(indexable)
+      index_single(indexable)
+
+      updated_entries = @files_to_entries[indexable.full_path]
+
+      return unless original_entries && updated_entries
+
+      # A change in one ancestor may impact several different others, which could be including that ancestor through
+      # indirect means like including a module that than includes the ancestor. Trying to figure out exactly which
+      # ancestors need to be deleted is too expensive. Therefore, if any of the namespace entries has a change to their
+      # ancestor hash, we clear all ancestors and start linearizing lazily again from scratch
+      original_map = T.cast(
+        original_entries.select { |e| e.is_a?(Entry::Namespace) },
+        T::Array[Entry::Namespace],
+      ).to_h { |e| [e.name, e.ancestor_hash] }
+
+      updated_map = T.cast(
+        updated_entries.select { |e| e.is_a?(Entry::Namespace) },
+        T::Array[Entry::Namespace],
+      ).to_h { |e| [e.name, e.ancestor_hash] }
+
+      @ancestors.clear if original_map.any? { |name, hash| updated_map[name] != hash }
```

Comment:
I wonder if we actually need to clear the whole cache? Can't we just clear the entries for which their linearization contains an impacted name?

Let's say I have

```rb
class A; end
class B < A; end
class C; end
```

And then I change if to

```rb
class A < C; end
class B < A; end
class C; end
```

I only need to discard the linearizations for `A` and `B` right? `C` is not impacted.

On a small example this may not save much, but for a fully linearized index for a huge codebase, maybe being smarter can save a lot?

Reply:
Let's go with clearing the cache for now and see if it becomes a burden as we use the feature.

The good news is we should be able to GC most of the old linearization and reduce the server memory footprint 😅